mono/mini/mini-llvm.c

```diff
@@ -9757,10 +9891,33 @@ mono_llvm_emit_aot_module (const char *filename, const char *cu_name)
 				/* The types might not match because the caller might pass an rgctx */
 				if (lmethod && LLVMTypeOf (callee) == LLVMTypeOf (lmethod)) {
 					mono_llvm_replace_uses_of (callee, lmethod);
+
+					g_hash_table_insert (directly_called, lmethod, ji->data.method);
 					mono_aot_mark_unused_llvm_plt_entry (ji);
 				}
 			}
 		}
+
+		// After direct calls are all computed, specialize directly called
+		// methods and propagate attrs from call sites
+		LLVMValueRef lmethod;
+		MonoMethod *method;
+		g_hash_table_iter_init (&iter, directly_called);
+		while (g_hash_table_iter_next (&iter, (void**)&lmethod, (void**)&method)) {
+			if (!mono_aot_can_specialize (method))
+				continue;
+
+			GArray *call_site_union = (GArray *) g_hash_table_lookup (module->method_to_call_info, method);
+
+			if (call_site_union)
+				g_assert (call_site_union->len == LLVMCountParams (lmethod));
+
+			for (int i = 0; call_site_union && i < call_site_union->len; i++)
+				if (g_array_index (call_site_union, gint32, i) == 0)
+					mono_llvm_propagate_nonnull_func (directly_called, module, lmethod, i);
+		}
```

Comment:
Is there any reason why you don't just queue up work items for every `lmethod, i` all at once and then kick off a single call to `mono_llvm_propagate_nonnull_func` that drains the queue?  Unless I'm missing something, `mono_llvm_propagate_nonnull_func` doesn't actually care about the identity of the root work item, right?

(I'm not sure if doing it that way would help/hinder anything.  So this comment might just be bikeshedding)

Reply:
@lambdageek the only reason I didn't build up an intermediate queue was that I wanted to avoid the allocation. The traversal pattern that we see above will be necessary either way if we want the same guarantee of termination (we can only decrement a nullable ref-count to zero once, after which it's dead). 

If we need to do the same traversal, building up the intermediate data structure on the C heap seemed like it would slow things down. This way, we have the queue-fill and queue-drain operations interleaved, keeping the work queue within reasonable (hopefully) bounds. 

If you think that joining the traversal and the action makes it a bit hard to read here, I can isolate the two into different functions without having to change allocation/control flow.

Reply:
I think I'll just move all of this logic into mono_llvm_propagate_nonnull_func, so the nested loops are easy to see